mirgecom/flux.py

```diff
@@ -35,6 +44,40 @@
 import numpy as np  # noqa
 
 
+def num_flux_lfr(f_minus, f_plus, q_minus, q_plus, lam, **kwargs):
+    """Lax-Friedrichs/Rusanov low-level numerical flux."""
+    return (f_minus + f_plus - lam*(q_plus - q_minus))/2
+
+
+def num_flux_central(f_minus, f_plus, **kwargs):
+    """Central low-level numerical flux."""
+    return (f_plus + f_minus)/2
+
+
+def hll_flux_driver(state_pair, physical_flux_func,
+                    s_minus, s_plus, normal):
+    """State-to-flux driver for hll numerical fluxes."""
+    actx = state_pair.int.array_context
+    zeros = 0.*state_pair.int.mass_density
+
+    f_minus = physical_flux_func(state_pair.int)@normal
+    f_plus = physical_flux_func(state_pair.ext)@normal
+    q_minus = state_pair.int.cv
+    q_plus = state_pair.ext.cv
+    f_star = (s_plus*f_minus - s_minus*f_plus
+              + s_plus*s_minus*(q_plus - q_minus))/(s_plus - s_minus)
+
+    # choose the correct f contribution based on the wave speeds
+    f_check_minus = actx.np.greater_equal(s_minus, zeros)*(0*f_minus + 1.0)
+    f_check_plus = actx.np.less_equal(s_plus, zeros)*(0*f_minus + 1.0)
```

Comment:
I understand why it was written like this (broadcasting issues with pyopencl, lazy (pytato), and numpy-like data structures), but I think we can probably simplify this once we pin down a good array context for testing (I'm thinking numpy is the way to go, and some work has started on it: https://github.com/inducer/arraycontext/pull/93).